src/inmanta/deploy/timers.py

```diff
@@ -0,0 +1,219 @@
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+    Contact: code@inmanta.com
+"""
+
+import asyncio
+import logging
+import sys
+from collections.abc import Coroutine
+from typing import Any, Callable
+
+import inmanta.deploy.scheduler
+from inmanta.agent import config as agent_config
+from inmanta.data.model import ResourceIdStr
+from inmanta.deploy.work import TaskPriority
+from inmanta.util import CronSchedule, ScheduledTask, Scheduler
+
+LOGGER = logging.getLogger(__name__)
+
+
+class ResourceTimer:
+    def __init__(self, resource: ResourceIdStr):
+        """
+        :param resource: The resource for which to ensure periodic repairs.
+        """
+        self.resource: ResourceIdStr = resource
+        self.time_to_next_deploy = sys.maxsize
+
+        self.repair_handle: asyncio.Task[None] | None = None
+        self.next_schedule_handle: asyncio.TimerHandle | None = None
+
+    def ensure_timer(
+        self,
+        periodic_deploy_interval: int | None,
+        periodic_repair_interval: int | None,
+        is_dirty: bool,
+        action_function: Callable[..., Coroutine[Any, Any, None]],
+    ) -> None:
+        """
+        Make sure the underlying resource is marked for execution
+
+        :param periodic_deploy_interval: Per-resource deploy interval, assumed to be a positive int or None
+        :param periodic_repair_interval: Per-resource repair interval, assumed to be a positive int or None
+        :param is_dirty: Last known state of the resource at the timer install request time.
+        :param action_function: The function to execute
+        """
+
+        next_execute_time: int
+
+        if periodic_repair_interval is None:
+            periodic_repair_interval = sys.maxsize
+        if periodic_deploy_interval is None:
+            periodic_deploy_interval = sys.maxsize
+
+        if is_dirty:
+            next_execute_time = min(periodic_deploy_interval, periodic_repair_interval)
+        else:
+            next_execute_time = periodic_repair_interval
+
+        if next_execute_time == sys.maxsize:
+            return
+
+        # We have to schedule the next deploy sooner than previously though:
+        # Cancel previous schedule and re-schedule
+        if next_execute_time < self.time_to_next_deploy:
+            self.cancel_next_schedule()
+
+        # A deploy is already scheduled for this resource
+        if self.next_schedule_handle is not None:
+            return
+
+        def _create_repair_task() -> None:
+            self.repair_handle = asyncio.create_task(action_function(self.resource, next_execute_time))
+
+        self.time_to_next_deploy = next_execute_time
+        self.next_schedule_handle = asyncio.get_running_loop().call_later(next_execute_time, _create_repair_task)
```

Comment:
shouldn't we have a different priorty/reason, based on what kind of trigger we use?